### PR TITLE
pallet-investments - Tests and Fixes

### LINKED ISSUE #917: Tests for `pallet-investments`
once #909 is closed we must implements unit tests for the pallet and its associated extrinsics and internal logic.

### PR DESCRIPTION
**Pull request by mustermeiszer**
# Description

Includes the unit tests for the pallet-investments. During testing I found a few bugs which are also fixed in the same PR. For easability of testing this also adds a Mock for the `trait InvestmentAccountant`

Closes #917 

## Changes and Descriptions
* Unit tests
* Bug fixes in investment logic
* Mock for `trait InvestmentAccountant`

## Type of change
- [x] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)

# How Has This Been Tested?
- [x] Unit test for pallet-investments

# Checklist:

- [x] I have added Rust doc comments to structs, enums, traits and functions
- [x] I have made corresponding changes to the documentation
- [x] I have performed a self-review of my code
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes
- [x] Any dependent changes have been merged and published in downstream modules
- [x] I rebased on the latest `parachain` branch

### CODE REVIEW

**Comment by mustermeiszer** [comment]
No worries, it's just a huge test file, plus a mock, plus some bug fixes.

File: libs/traits/src/lib.rs [hunk 1]
```diff
@@ -372,7 +372,7 @@ pub trait OrderManager {
 	) -> Result<(), Self::Error>;
 }
 
-/// A trait whos implementer provides means of accounting
+/// A trait who's implementer provides means of accounting
```
Comment (gruberb) [inline libs/traits/src/lib.rs:375]: It's actually `whose` 😄

File: libs/traits/src/mocks/accountant.rs [hunk 2]
```diff
@@ -0,0 +1,236 @@
...
+/// Exposes a struct $name that implements the `trait Accountant`. The struct expects one generic
+/// parameter that implements the fungibles traits `Inspect`, `Mutate` and `Transfer`. Furthermore,
+/// there exists a struct `GenesisConfig` that implements `trait GenesisBuild` that can be used
+/// like any other `GenesisConfig` to initialize state in the `TestExternalities`.
+///
+/// Also exports a `struct InvestmentInfo` to be used in the `GenesisConfig`
+///
+/// * E.g.: `MockAccountant<Tokens: frame_support::traits::tokens::fungibles::{Inspect, Mutate, Transfer}>`
+///
+/// # Example macro usage:
```
Comment (gruberb) [inline libs/traits/src/mocks/accountant.rs:22]: Nice!!

File: pallets/investments/src/lib.rs [hunk 3]
```diff
@@ -348,38 +357,66 @@ pub mod pallet {
...
+			who: T::AccountId,
+			investment_id: T::InvestmentId,
+		},
 	}
 
 	// Errors inform users that something went wrong.
 	#[pallet::error]
 	pub enum Error<T> {
-		/// The order has been marked as cleared. It's either active or
+		/// The order has not been marked as cleared. It's either active or
```
Comment (gruberb) [inline pallets/investments/src/lib.rs:413]: 👍

File: pallets/investments/src/lib.rs [hunk 4]
```diff
@@ -655,204 +649,244 @@ where
...
+		} else {
+			// In this case the user has no active position.
+			// We remove the order from the system then.
+			// This is also important in cases where all redemptions
+			// where fulfilled before the order reached the current OrderId
+			// counter. Users can in this case now simply submit a new order.
+			true
+		};
+
+		if remove_user_order {
```
Comment (gruberb) [inline pallets/investments/src/lib.rs:698]: Why not move this into the `else` block and remove `remove_user_order`?

**Review by gruberb — CHANGES_REQUESTED** [review]
Beautiful 💐

I need some more time to actually understand what's going on, so I won't hold back approving. There is just one little comment where I think we can move something inside an `else` block and get rid of a variable, otherwise very nice 👍

**Comment by mustermeiszer** [comment]
@gruberb addressed

**Review by gruberb — APPROVED** [review]
LGTM 👍

**Review by gruberb — APPROVED** [review]
🚀

**Pull request merged**